### ISSUE TITLE
Session Garbage collection fails for RedisBackend if data is not in sync

### ISSUE THREAD

**Issue by DavidSporer** [neos/flow-development-collection#2182]
### Description
This may happen rarely but it did in our case ;-)
When using the RedisBackend, the garbage collection for sessions works in a way that it first checks if the iterator is valid. If that's not the case it will simply skip.
In general there's an entry in Redis called "entries" (e.g. Flow_Session_MetaData:entries) that contains all keys that supposed to exist (e.g. in MetaData cache). Now if - for whatever reason, we haven't figured that out - the first entry in "entries" does not exist anymore the iterator is invalid (also the backend's valid-function itself returns false in this case) which means garbage collection won't happen anymore from that point onwards.
This can lead to a huge list of entries and session data never being removed.

### Steps to Reproduce

@robertlemke and I worked on this together - maybe he can amend on that.
1. Do something in Flow that creates a session, e.g. logging in.
2. Remove the entry of that session from MetaData cache but leave it in Flow_Session_MetaData:entries and ensure that this entry is the first one in that list.
3. Garbage collection will immediately skip even if there's millions of other entries in Flow_Session_MetaData:entries that should be removed.

#### Expected behavior

I'd argue that there should be some sort of cleanup mechanism that removes orphaned entries or - if possible - invalid entries are ignored.
One way would be to include such a cleanup mechanism in the garbage collection and skip cleanup after a couple of entries so that it won't block the application for too long but eventually starts working again. In our case we added a command that does this cleanup task for more than half a million entries which took an hour approximately.

This is the command (thanks @robertlemke ):
`
/**
     * Fix session metadata
     *
     * @param string $cacheIdentifier
     *
     * @throws Exception
     * @throws NoSuchCacheException
     */
    public function fixRedisSessionMetadataCommand($cacheIdentifier = 'Flow_Session_MetaData'): void {
        $cache = $this->cacheManager->getCache($cacheIdentifier);
        $cacheBackend = $cache->getBackend();

        assert($cacheBackend instanceof RedisBackend);

        $cacheConfigurations = $this->cacheManager->getCacheConfigurations();

        $redis = new \Redis();

        $hostname = $cacheConfigurations[$cacheIdentifier]['backendOptions']['hostname'];
        $password = $cacheConfigurations[$cacheIdentifier]['backendOptions']['password'];
        $database = $cacheConfigurations[$cacheIdentifier]['backendOptions']['database'];

        try {
            if (strpos($hostname, '/') !== false) {
                $connected = $redis->connect($hostname);
            } else {
                $connected = $redis->connect($hostname, 6379);
            }
        } finally {
            if ($connected === false) {
                throw new Exception('Could not connect to Redis.', 1391972021);
            }
        }

        if ($password !== '') {
            if (!$redis->auth($password)) {
                throw new Exception('Redis authentication failed.', 1502366200);
            }
        }
        $redis->select($database);

        $cacheBackend->rewind();
        $valid = $cacheBackend->valid();
        $this->outputLine('Backend iterator is ' . ($valid ? '<success>valid</success>' : '<error>invalid</error>'));

        $entriesCount = $redis->lLen($cacheIdentifier . ':entries');
        $this->outputLine('Entries list contains %s entries', [$entriesCount]);

        $this->outputLine();
        $this->outputLine('Checking if entries exist for all keys in "entries" ...');
        $this->outputLine();

        $knownKeys = $redis->lRange($cacheIdentifier . ':entries', 0, $entriesCount);
        foreach ($knownKeys as $knownKey) {
            $keyWithPrefix = $cacheIdentifier . ':entry:' . $knownKey;
            if ($redis->exists($keyWithPrefix)) {
                continue;
            }
            $this->outputLine('<error>Removing key %s</error>', [$keyWithPrefix]);
            $redis->lRem($cacheIdentifier . ':entries', $knownKey, 0);
        }

        $this->outputLine();
        $this->outputLine('Checking if entries exist which are not listed in "entries" ...');
        $this->outputLine();

        $existingKeys = $redis->keys($cacheIdentifier . ':entry:*');
        $keysCount = count($existingKeys);
        $this->outputLine('Found %s keys with prefix %s', [$keysCount, $cacheIdentifier . ':entry:*']);

        foreach ($existingKeys as $existingKeyWithPrefix) {
            $existingKey = str_replace($cacheIdentifier . ':entry:', '', $existingKeyWithPrefix);
            if (in_array($existingKey, $knownKeys, true)) {
                continue;
            }
            $this->outputLine('<error>Removing entry with key %s</error>', [$existingKey]);

            $redis->del($existingKeyWithPrefix);
        }
    }
`

I'll be happy to work on a PR for this but wanted to discuss what's the expected behaviour first. I think this is clearly a bug because an invalid entry - no matter where it came from - shouldn't stop the garbage collection from working.

#### Actual behavior

RedisBackend's valid-function (and the iterator as well) become invalid resulting in a broken garbage collection

### Affected Versions

Flow: we've tested with Flow 6.3.1 but I assume this is in there since the RedisBackend exists.

**Comment by albe** [comment]
Interesting find! I'm not so deep into the redis cache backend, but this sounds fixable?

**Comment by sorenmalling** [comment]
From your reproduction steps

> Remove the entry of that session from MetaData cache but leave it in Flow_Session_MetaData:entries and ensure that this entry is the first one in that list.

When will it be removed from one, and not the other? Did you manage to pinpoint, how/where/why this occured?

**Comment by DavidSporer** [comment]
> When will it be removed from one, and not the other? Did you manage to pinpoint, how/where/why this occured?

Not really. Our best guess was that Redis got restarted at some point and had corrupt data afterwards. This should be an edge case but one that can happen. And I’m our case it lead to Flow constantly „forgetting“ about sessions.

**Comment by sorenmalling** [comment]
Okay, got it 👍 Can you be so kind to try and adopt the commandcontroller functionality into the garbage collection of the Redis implementation?

**Comment by DavidSporer** [comment]
Yes, I'll try to do that in the next few weeks.
We don't seem to be the only ones experiencing that issue. Comment from Slack:
"_Aaaahhh... Damn! We had been there too. The problem can arise if you have expering keys (they never get removed from "entries" list, but this only make the entries list bigger) or you allow entries to be evicted. Eviction might hit the entry list. From that point on, Flow will never clean up the left over keys.
At least that is what happened to us. So we disabled automatic eviction. And do some manual clean-up in the redis cache from time to time. (another problem that we experienced is, that the flush operation might take so long, that the sentinels decide to initiate a roll over - which me managed with a non blocking "flushByIteration" - which takes significantly more time, but does not lock the system and most important - does not trigger a failover)_"

**Comment by kdambekalns** [comment]
Could that be somehow related to https://github.com/neos/flow-development-collection/pull/2052?

**Comment by DavidSporer** [comment]
It sounds as if the PR would not necessarily solve the exact issue we had because data was corrupt in our case.
But if entries would get removed after a certain amount of time automatically (because of the TTL), it would eventually lead to a state where our problem would have been fixed automatically.
If the #2052 works reliably you wouldn't even need a garbage collection anymore I assume?

**Comment by kaystrobach** [comment]
For what reason do we have the entries key? Kmho this is a bad design. As redis keys * provides the info.

**Comment by kaystrobach** [comment]
Have you tried : https://github.com/sandstorm/OptimizedRedisCacheBackend

**Comment by kaystrobach** [comment]
it completely removes the usage of entries and should be the better solution imho ...
